### PR TITLE
haku: dedicated 2-layer MLP volume decoder head (vol_pressure 2.08x gap)

### PR DESCRIPTION
**Pull request by morganmcg1**
# Dedicated Volume Decoder Head (2-layer MLP vs single Linear)

## Hypothesis

The current `SurfaceTransolver` uses a single `LinearProjection` (a bare `nn.Linear`) as the
volume output head:

```python
self.volume_out = LinearProjection(n_hidden, self.volume_output_dim)
```

This maps the 512-d backbone output directly to volume pressure in one linear step. Surface tokens
get the same treatment via `self.surface_out`. Both surface and volume targets are solved jointly
in the backbone, but the readout capacity is symmetric even though the volume field (`vol_pressure`)
shows the largest gap vs AB-UPT (2.08×: 12.656% vs 6.08% reference). A deeper, non-linear
decoder head for volume could provide the additional expressive capacity needed to resolve
the volume pressure field from the shared backbone features.

**Prediction**: Replacing `volume_out = LinearProjection(n_hidden, n_hidden)` with a 2-layer MLP
(n_hidden → n_hidden → vol_output_dim with GELU) will reduce vol_pressure error and drag down
val_abupt overall.

## Current SOTA Baseline (PR #232, askeladd, run `r8s2dtnq`)

| Metric                 | val       | test      |
|------------------------|-----------|-----------|
| abupt_axis_mean_rel_l2 | **9.065%** | 10.190%  |
| tau_y                  | 11.952%   | —         |
| tau_z                  | 12.447%   | —         |
| vol_pressure           | 12.656%   | —         |

SOTA recipe: `--optimizer lion --lr 1e-4 --weight-decay 5e-4 --no-compile --batch-size 4 --ema-decay 0.999 --lr-warmup-epochs 1 --model-layers 4 --model-hidden-dim 512 --model-heads 4 --model-slices 128 --model-mlp-ratio 4`

Beat target: val_abupt < **9.065%**.

## Code Changes Required

You must modify **`target/model.py`** and **`target/train.py`**. No other files.

### 1. Add `model_vol_decoder_depth` to `Config` in `train.py`

In the `Config` dataclass (around line 88), add after `model_dropout`:

```python
model_vol_decoder_depth: int = 1   # 1=linear (baseline), 2=MLP decoder
```

This auto-registers as `--model-vol-decoder-depth INT` CLI flag. Default of 1 preserves the
existing behaviour exactly so you can run baseline and experiment from the same codebase.

### 2. Pass `vol_decoder_depth` to `SurfaceTransolver` in `train.py`

In `build_model()` (around line 170), add:

```python
def build_model(config: Config) -> SurfaceTransolver:
    return SurfaceTransolver(
        n_layers=config.model_layers,
        n_hidden=config.model_hidden_dim,
        dropout=config.model_dropout,
        n_head=config.model_heads,
        mlp_ratio=config.model_mlp_ratio,
        slice_num=config.model_slices,
        rff_num_features=config.rff_num_features,
        rff_sigma=config.rff_sigma,
        vol_decoder_depth=config.model_vol_decoder_depth,   # <-- add this
    )
```

### 3. Replace `volume_out` in `SurfaceTransolver.__init__` in `model.py`

Add `vol_decoder_depth: int = 1` to `SurfaceTransolver.__init__`'s signature.

Then replace the fixed linear head with a depth-conditional head:

```python
# Replace this line:
#   self.volume_out = LinearProjection(n_hidden, self.volume_output_dim)
# With:
if vol_decoder_depth >= 2:
    self.volume_out = nn.Sequential(
        nn.Linear(n_hidden, n_hidden),
        nn.GELU(),
        nn.LayerNorm(n_hidden, eps=1e-6),
        nn.Linear(n_hidden, self.volume_output_dim),
    )
else:
    self.volume_out = LinearProjection(n_hidden, self.volume_output_dim)
```

No changes needed to `SurfaceTransolver.forward()` — `self.volume_out(volume_hidden)` works
for both `nn.Sequential` and `LinearProjection`.

## Experiment: Two-arm DDP8 sweep

Run both arms with `--wandb-group haku-vol-decoder` so runs are grouped:

**Arm A — 2-layer MLP volume decoder (experimental)**

```bash
torchrun --standalone --nproc_per_node=8 train.py \
  --wandb-group haku-vol-decoder \
  --wandb-name haku-vol-decoder-depth2 \
  --optimizer lion \
  --lr 1e-4 \
  --weight-decay 5e-4 \
  --no-compile \
  --batch-size 4 \
  --ema-decay 0.999 \
  --lr-warmup-epochs 1 \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 4 \
  --model-slices 128 \
  --model-mlp-ratio 4 \
  --model-vol-decoder-depth 2
```

**Arm B — Baseline control (linear head, depth=1)**

```bash
torchrun --standalone --nproc_per_node=8 train.py \
  --wandb-group haku-vol-decoder \
  --wandb-name haku-vol-decoder-depth1-ctrl \
  --optimizer lion \
  --lr 1e-4 \
  --weight-decay 5e-4 \
  --no-compile \
  --batch-size 4 \
  --ema-decay 0.999 \
  --lr-warmup-epochs 1 \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 4 \
  --model-slices 128 \
  --model-mlp-ratio 4 \
  --model-vol-decoder-depth 1
```

Run Arm A and Arm B sequentially (single pod with 8 GPUs, one run at a time).

## Reporting

Post results as a PR comment with this table:

| Arm | vol_decoder_depth | val_abupt | tau_y | tau_z | vol_pressure | W&B run |
|-----|-------------------|-----------|-------|-------|--------------|---------|
| A   | 2 (MLP)           | …         | …     | …     | …            | …       |
| B   | 1 (linear ctrl)   | …         | …     | …     | …            | …       |
| SOTA baseline | 1 (linear) | 9.065% | 11.952% | 12.447% | 12.656% | r8s2dtnq |

Key question: does vol_pressure improve in Arm A vs the control, and does this translate
to overall val_abupt improvement?

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #321 because ADVISOR: Closing as orphaned. Student 'haku' is not currently allocated to the tay branch (no senpai-drivaerml-ddp8-haku deployment exists). This PR was created in a prior advisor session by mistake. The volume-decoder-head hypothesis is well-designed and being reassigned to tanjiro (whose previous PR #300 sandwich-norm just closed as a NEGATIVE result)..
